### ISSUE TITLE
Improve support for Macbook Pro TouchID 

### ISSUE THREAD

**Issue by mxk6n** [keepassxreboot/keepassxc#2865]
To further improve the utilization of TouchID on Macbook Pro and make the unlock feature more intuitive to use (see comments in #2720 for pull request #1851), I would like to implement the following changes:

- Make enabling TouchID a **global setting** (not per database)
- Add an **indication that TouchID is available to unlock** (colored icon) on the unlock screen / greyed out when it is not. When it is disabled in the settings no icon is shown.
- Unlock with **TouchID is triggered** when it is activated/available and ...
  - ... User *clicks* on the colored icon
  - ... User tries to unlock database *without entering a password*
  - ... Window is being *brought to the foreground* by the user
  - ... *Tab is being changed* / brought to foreground (multi database usage)
  - I would suggest making the last two scenarios an optional setting
- TouchID is automatically **deactivated on 3 consecutively failed tries** (until user enters the password correctly again)
- **Raising the maximum time-out** of TouchID (currently the maximum value is 999 minutes)

Further additions (edited):

- **Hints and dialog** that guide the user through the process of activating and using TouchID for Keepass
- **Enable** TouchID persistently **across restarts** of the application

How do you guys feel about this?

**Comment by marco-brandizi** [comment]
I took a while to understand that I have to both tickbox the touchID option AND click on OK without typing any password. I don't get if this enhancement issue is taking this into account, to me the procedure should be clearer, eg, just start polling the TouchID sensor and unlock when it sends back a positive fingerprint recognition. Alternatively, place a button like "Unlock with TouchID now".

**Comment by mxk6n** [comment]
Yes exactly, I totally agree with you! I suggested to get rid of the checkbox (just one global setting for disabling TouchID altogether in the settings menu). Also adding a button / clickable biometric icon for unlocking with TouchID manually.

Unfortunately, it is not possible (to my knowledge) on macOS to poll the fingerprint sensor in the background - it has to be triggered, which then brings this system popup to the foreground. Therefore, I made the suggestion to automatically do this when the application is being brought to the foreground (btw. that is also how 1pw does it).

**Comment by hhrutter** [comment]
This does not work for me on 2.4.0 and MacOS 10.14.4.
I get `Unable to open the database` when I check `TouchId for quick unlock` and leave the password field blank.

How is that checkbox supposed to work?

**Comment by mxk6n** [comment]
@hhrutter This is just the thread for the upcoming improvements - non of these are implemented yet (I am currently working on it, though).

As for your problem, you have to have both `Password` and `TouchID` checked when unlocking with your password for the first time in order to **activate** TouchID. For subsequent unlock attempts you can then leave the password field empty, press `Unlock` which will then ask for your fingerprint to unlock the database.

**Comment by hhrutter** [comment]
When are you supposed to get an `Unlock` button?
When I start KeePassXC I have to provide my keyfile and password and check `TouchId for quick unlock`? 

Ok. I am in.

This feature is only for unlocking after the db locks due to  timeout or manual lock?
Even if I get this to work - which I don't because I don't get an `Unlock` button/option just the initial login form where I have to provide all my credentials - can't we have this checkbox for acrivation TouchId for any future login when starting up KeepassXC?

KeepassXC rocks but this is frustrating I have to admit.

**Comment by droidmonkey** [comment]
@hhrutter sorry for all the ux issues with this. Unfortunately the lead devs do not have a mac with touchid hardware so we could not fully test this feature. Luckily mxk6n is planning to upgrade it.

**Comment by natemo4873** [comment]
This thread helped me understand how to use it. its a shame i cant use touch id until after an initial master password entry. but its still handy though. thanks for all the hard work. Love your guys password manager app!

**Comment by mxk6n** [comment]
Sorry for the confusion about this... I will definitely add some explanations / dialogs to guide through the process of activating and using TouchID together with the improvements mentioned above.

In order to activate TouchID we will always need the initial entry of the master password. However, right now this is true for every restart of the application (as the encrypted secret is kept in memory which is cleared when closing Keepass). You raise a good point, I will try to facilitate the Secure Enclave to perform the decryption of a persistent encrypted secret which will then survive the restart of the application :) I added it as a bullet in my initial comment.

**Comment by aserrallerios** [comment]
In 2.4.1 and I still cannot see the `Unlock` button when I log in and then lock the DB.

There's something I could be missing?

**Comment by mxk6n** [comment]
Just press `ok`, that's what i meant with unlock ;)

**Comment by hhrutter** [comment]
Hooray!!! Managed to unlock via TouchId for the very first time.

I would suggest having a Unlock via TouchId or Let me in button. Having to
say OK is not intuitive in this situation.

**Comment by mkozjak** [comment]
Hi! This is not a rant. Just a honest feeling about this feature.
I literally used it zero times so far, even though I wanted to use it. I suppose my laptop use case doesn't apply to this feature. I open and close the lid multiple times a day with keepass ending up locked, which is nice. But I never got into a situation where I'd be able to use Touch ID without my master password. tldr; Not sure if it makes sense, but for me it'd be cool to have Touch ID without the password over a big period of time, eg. a few days. Similar to something LG does with their fingerprint sensor on the back. They demand a password retype every 48 hours. But between those, you can just use your fingerprint. Since I open the app at least 5 times a day, maybe that would make sense?..

**Comment by sinistersnare** [comment]
Hi, I see a lot of negativity and I just want to say that I love this feature. It is so nice to have this 'quick unlock' like feature available that is still secure (at least secure enough).

Thanks for implementing it! Although understanding exactly how it works at first was a bit difficult (so work can still be done), I think its awesome.

**Comment by johnrichardrinehart** [comment]
What is the progress on changing the UI so implement a single button to trigger Touch ID authentication? I also didn't know that I had to click `OK` *and* check the box until I [saw @marco-brandizi's response](https://github.com/keepassxreboot/keepassxc/issues/2865#issuecomment-478579122).

**Comment by tomhundt** [comment]
I'm okay with having to click a button to trigger TouchID.  _Especially_ if that button has a keyboard shortcut! 😄   **Which, at the moment, seems to be Cmd+Enter** 👍

**Comment by Agentscreech** [comment]
I'm having trouble getting the touch id to work for this.  Here's the workflow

- Launch app
- select database
- enter password while having the TouchID for Quick Unlock box checked
- hit enter
- Log in will be successful
- manually click lock database
- back at the unlock screen, I click the TouchID for Quick Unlock button
- click ok

Expected: TouchID prompt of some kind

Observed: "Unlocking the database failed and you did not enter a password"

What am I missing?

**Comment by johnrichardrinehart** [comment]
@Agentscreech Can you confirm your KeepassXC and OS versions? I'll try to reproduce.

**Comment by Agentscreech** [comment]
Thanks for the reply. KeePassXC v2.5.1 on macOS 10.14.6

**Comment by johnrichardrinehart** [comment]
@Agentscreech I am on macOS 10.15.1 (Catalina) and the same version of KeePassXC.

Following your steps above *does* result in successful authentication via TouchID. So, I'm unable to reproduce your issue.

However, I don't recommend you update to Catalina to support this feature. There is likely (read: hopefully) another cause for the discrepancy.

**Comment by Agentscreech** [comment]
Thanks.  Is there something I can reference about where I should be looking for more troubleshooting?  I feel like I followed the build instructions correctly. FYI, touchID wasn't setup (new computer) when I built it.

**Comment by johnrichardrinehart** [comment]
Did you build from source or install from disk image? I'd recommend trying to install from the [provided disk image](https://github.com/keepassxreboot/keepassxc/releases/download/2.5.1/KeePassXC-2.5.1.dmg) first.

**Comment by Agentscreech** [comment]
That was it.  Must of borked something when I tried to build it.  I was looking for the pre-built image, but missed it somehow.  Thanks for the link!

**Comment by taelerwatkins-tcn** [comment]
I'm not seeing any TouchID UI on 2.5.2. It says it's enabled in the debug information, but there doesn't seem to be a way to activate it. Did something change? Just using the regular build. Didn't do anything funky on my end.

```
KeePassXC - Version 2.5.2
Revision: 62cda9d

Qt 5.14.0
Debugging mode is disabled.

Operating system: macOS 10.15
CPU architecture: x86_64
Kernel: darwin 19.0.0

Enabled extensions:
- Auto-Type
- Browser Integration
- SSH Agent
- KeeShare (signed and unsigned sharing)
- YubiKey
- TouchID

Cryptographic libraries:
 libgcrypt 1.8.5
```
<img width="236" alt="Screen Shot 2020-01-07 at 2 21 33 PM" src="https://user-images.githubusercontent.com/59576782/71930517-0ee59a00-3159-11ea-9851-0caee9a19b56.png">

**Comment by johnrichardrinehart** [comment]
![image](https://user-images.githubusercontent.com/6321578/71934969-7ea85480-315a-11ea-9477-4f1ce4686dbe.png)

@taelerwatkins-tcn Do you see the checkbox circled in yellow? That's the UI element which enables/disables TouchID for KeePassXC.

**Comment by droidmonkey** [comment]
@johnrichardrinehart you need to update your KeePassXC!

**Comment by johnrichardrinehart** [comment]
@droidmonkey Thank you! I was on 2.4.x.

@taelerwatkins-tcn You should instead see a similar UI element as follows in 2.5.2

![image](https://user-images.githubusercontent.com/6321578/71935736-70f3ce80-315c-11ea-8ec3-aba38737e687.png)

**Comment by taelerwatkins-tcn** [comment]
@johnrichardrinehart Nada. This is a completely new install on a new machine that was just set up yesterday. Been unlocking manually a ton now and discovered this mac has TouchID and eventually landed here when I wasn't able to use it.

<img width="1003" alt="Screen Shot 2020-01-07 at 4 11 53 PM" src="https://user-images.githubusercontent.com/59576782/71937119-a1da0080-3168-11ea-8d55-fe2b78464e47.png">

**Comment by johnrichardrinehart** [comment]
@taelerwatkins-tcn Can you confirm the version of KeePassXC?

**Comment by taelerwatkins-tcn** [comment]
@johnrichardrinehart 
<img width="1003" alt="Screen Shot 2020-01-07 at 4 36 16 PM" src="https://user-images.githubusercontent.com/59576782/71938247-0a76ac80-316c-11ea-9b86-ecce33385ee1.png">

**Comment by foufoulefou** [comment]
Hello, 

I have the same problem with a touch ID who doesn't work. And as it say here I've check the touch ID box before entering my password.

I have a macbook pro 16" and here's my config in keepassXC :

KeePassXC - Version 2.5.2
Revision: 62cda9d

Qt 5.14.0
Debugging mode is disabled.

Operating system: macOS 10.15
CPU architecture: x86_64
Kernel: darwin 19.2.0

Enabled extensions:
- Auto-Type
- Browser Integration
- SSH Agent
- KeeShare (signed and unsigned sharing)
- YubiKey
- TouchID

Cryptographic libraries:
 libgcrypt 1.8.5

If you have a clue don't hesitate, I double check every step said here and it won't work.
Thanks

**Comment by johnrichardrinehart** [comment]
@foufoulefou Did you compile from source? Try installing an [official release](https://github.com/keepassxreboot/keepassxc/releases/download/2.5.3/KeePassXC-2.5.3-Sierra.dmg) and see what happens.

**Comment by taelerwatkins-tcn** [comment]
@foufoulefou just double checking, but here's the basic steps.

1: Set up TouchID for the machine
2: Open database and unlock normally with the TouchID checkbox checked.
3: Lock database.
4: From lock screen, leave password blank and click OK while TouchID checkbox is checked. TouchID should show a dialog that KeePassXC is requesting TouchID authentication.
5: Touch to unlock.

*NOTE* the standard settings make you authenticate normally if the screen locks. There is also a setting you can enable to have a configurable timeout, after which you are required to authenticate normally.

**Comment by foufoulefou** [comment]
@johnrichardrinehart yes I have tried with an official release, and even with the one you give., but the answer was that I didn't inderstood a thing.

Thing that @taelerwatkins-tcn said help me understand, when I try to use touch ID, I never try to click OK to make the touch ID box appear. Now that I did it, everything is going well.
So thank you to you both, now it work well :)

**Comment by taelerwatkins-tcn** [comment]
@foufoulefou Glad to help.

I feel that the UI doesn't self-document how it's meant to be used. I'm not a UI person, so I don't have any real suggestions for making it more understandable.

**Comment by leviem1** [comment]
@mxk6n Maybe add a button that says "Unlock with TouchID" instead of the checkbox? That way it's more of an active choice than a passive option.

**Comment by taelerwatkins-tcn** [comment]
@leviem1 That would make more sense. Then when initializing it or after it times out etc, it would just function as it does now in showing the message you need to relogin and act as a login button that, when used with regular authentication, would activate it for TouchID until timeout again.

Basically, a specific secondary login button vs a checkbox that either uses TouchID if already active, or activates it upon logging in using that button.

The main issue is that there needs to be an action to ask for TouchID. It can't just wait for a fingerprint like it does on Android, which thinking now is convenient, but I'm wondering if the sensor for Android shouldn't have the same sort of restrictions placed on it by the OS. But that is irrelevant for this discussion.

**Comment by johnrichardrinehart** [comment]
> The main issue is that there needs to be an action to ask for TouchID. It can't just wait for a fingerprint like it does on Android

Hmm, maybe I'm missing something, but the only time that KeePassXC needs to "ask for TouchID" is when both: 1) TouchID has timed out and 2) the user has pressed the "Unlock with TouchID" button. Then, it seems like KeePassXC can raise a prompt similar to the below (sorry for the pixelation... too lazy to fix).

<img width="448" alt="Screen Shot 2020-01-27 at 11 07 57 AM" src="https://user-images.githubusercontent.com/6321578/73205213-6a220100-40f5-11ea-98d3-f9998c569513.png">

**Comment by taelerwatkins-tcn** [comment]
@johnrichardrinehart Either I misspoke, or you misread. My apologies. Let me clarify. My meaning was that the Android app does not have this sort of weird UI issue because if the app is open at the unlock screen, it's automatically waiting for a fingeprint, so there doesn't need to be any UI to click or any confusion like that. macos enforces that you cannot poll for the fingerprint sensor without a deliberate user action (which opens the dialog you reference), which is why you have to click "OK" in order to get the prompt, as opposed to just being able to touch the sensor.

That is all I meant. Was just drawing the parallel between platforms, and elaborating as to why the TouchID UI isn't quite so straightforward because it does require a deliberate click of some sort of button to activate.

Also, sorry to drag the issue down a tangent.

**Comment by johnrichardrinehart** [comment]
@taelerwatkins-tcn That's perfectly clear. Thank you.

I think allowing TouchID only after first log-in using password (and/or key file) is good security policy. This policy bridges the design gap between needing to type a (potentially very long) password many times and allowing anyone with a facsimile of the user's fingerprint access to the machine.

For instance, macOS allows for a logged out user to log in from shutdown only by using the password. On re-log-in (without system reset) the user is eligible to use TouchID to authenticate. I think this is fine.

**Comment by leviem1** [comment]
@johnrichardrinehart @taelerwatkins-tcn Agreed, I think the general flow is familiar enough to be intuitive, there just needs to be better visual elements. I think it would be simple enough to have a button that is specifically for TouchID that only appears when using TouchID is available (i.e. after the initial login with a password). That way users have a clear distinction of when and how to use it.

**Comment by gregfenton** [comment]
I am new to Mac (new MBP) ; been using Keepass on Windows and Android for years.

KeePassDroid on my Samsung Galaxy allows me to open the .kdbx file with only my fingerprint.  To set it up, you enter your password on the Open screen and instead of pressing OK you put your finger on the fingerprint reader.  If the password is correct, the file opens and it gets securely stored against your fingerprint.  Subsequent opening of that same .kdbx can be done with just your fingerprint - no need to enter the password ever again.

I'd like to see similar in KeepassXC on the MBP.  Happy to help out where I can.  I have dabbled a bit in XCode development for an iPad app, so willing to try things out if someone is willing to direct me

Two things I'd like to see in KeepassXC:

1. TouchID unlock that once setup "works forever" (or until the config gets wiped from the keystore)
2. Update the KeepassXC documentation to reflect the TouchID functionality (I don't see it listed in the github overview, nor the docs/FAQ on https://keepassxc.org/

**Comment by tbleich** [comment]
Touch ID unlock also not working on my 16" MBP (Catalina 10.15.3). 
I do not get prompted for TouchID when the checkbox "TouchID for quick unlock" is checked and the database was unlocked by password before. 

Maybe its a pattern, 16" MBP?

My config:
```
KeePassXC - Version 2.5.3
Revision: f8c962b

Qt 5.14.0
Diagnosemodus ist deaktiviert.

Betriebssystem: macOS 10.15
CPU-Architektur: x86_64
Kernel: darwin 19.3.0

Aktivierte Erweiterungen:
- Auto-Type
- Browser-Integration
- SSH-Agent
- KeeShare (bestätigtes und unbestätigtes Teilen)
- YubiKey
- TouchID

Kryptographische Bibliotheken:
 libgcrypt 1.8.5
```

**Comment by RayfenWindspear** [comment]
> Touch ID unlock also not working on my 16" MBP (Catalina 10.15.3).
> I do not get prompted for TouchID when the checkbox "TouchID for quick unlock" is checked and the database was unlocked by password before.

Did you change any of the TouchID settings in KeePass? The defaults are for it to require password anytime your screen locks, and there is a setting disabled by default where you can set a timeout which will require password again after `x` minutes. 

I suspect the issue getting you is the screen lock setting. Unlock, and in settings you can disable that functionality if desired. With both disabled, the only time you should need to use password is if you close the application. 

Oh, that's the other thing. TouchID only stays active if you DO NOT close KeePassXC. If you quit, you will need password again.

**Comment by gregfenton** [comment]
> @ tbleich
> Did you change any of the TouchID settings in KeePass? The defaults are for it to require password anytime your screen locks, and there is a setting disabled by default where you can set a timeout which will require password again after `x` minutes.
> 

**KeePassXC >> Preferences... >> Security >>** (uncheck) **Forget TouchID when session is locked or lid is closed**

Thanks for pointing this out!  I have now unchecked the above setting and the touch ID works for unlocking for me!   I locked the database (clicked the Lock icon in the toolbar), then clicked the **OK** button on the Unlock KeePassXC Database screen and I was prompted for my TouchID.

As my entry above suggests: a bit of updating of documentation could go a long way here.  I'll submit a push request shortly.

**Comment by RayfenWindspear** [comment]
I concur. It's a little confusing how it works and why it isn't working when you think it should. It's a UI and/or documentation disconnect with the user's perception of it's functionality.

**Comment by tbleich** [comment]
@RayfenWindspear 
I did not close KeePassXC when not using it and I followed the instructions of how to use TouchID. I am not sure whether I changed some settings and will check the recommendation of @gregfenton to make it work.

Unfortunately a colleague spilled a bottle of water over my laptop, the repair takes about three weeks and until then I am on a TouchID-less Macbook...

**Comment by droidmonkey** [comment]
I hope they at least bought you lunch!

**Comment by tbleich** [comment]
@gregfenton @RayfenWindspear Thanks for the hint, I had the box checked. Unchecked it, works like a charm now! 

**KeePassXC >> Preferences... >> Security >> (uncheck) Forget TouchID when session is locked or lid is closed**

Yes, I got lunch :) In case you are wondering, the price for a new 16" MBP display is around 600€ net, excluding labor ...

**Comment by foufoulefou** [comment]
Hello, 

Since I updated my os to mac os 10.15.4, my keepass version didn't launch, so I had to re-install it.
After that, keepass ( version 2.5.3 ) launched again correctly but it's touch ID which is not working now.

To resume what happened, here is every step
- launch keepaas 
- check the box "unlock with touch ID"
- entering my password by typing it, then press enter
- locking my session again by clicking on the lock 

And the problem appears here, the "unlock with touch ID" is uncheck at this point whereas I uncheck the box "forget touch ID when session is lock" in the preferences -> security option.

Despite this, I try to unlock with touch ID, so I checked the box first, then I pressed enter and this window appeared.
![window](https://user-images.githubusercontent.com/60021561/77667076-0b0c2b80-6f82-11ea-965f-e7adfbeb2ca2.png)
I tried to go in "DatabaseSettings/Security" as said in the window, but I couldn't find an option to reset my password.

If you have any ideas about this problem, let me know.
Thanks

**Comment by droidmonkey** [comment]
We know. This was the trade-off required to get an emergency patch out there to keep the application running.

**Comment by foufoulefou** [comment]
Ok thank, I thought I  did something wrong. I will wait for a patch. 

Thanks

**Comment by edswangren** [comment]
As a long time user, just throwing in my $0.02...

I am also in favor of simplifying this feature. If I've enabled TouchId, here's what I would like the flow to look like:

1. Open DB or pull up window of a timed out DB
2. Click a button
3. Provide fingerprint

That's it. I'd like to just be able to use my FP at all times on my personal PC. For the average user it makes no sense to complicate things further.

**Comment by phoerious** [comment]
Using TouchID effectively lowers the security of your database. It is still secure against theft from your cloud storage, but you trade a complex master password for your arguably complex, but easily obtainable yet unchangeable finger print. I am not against convenience features and for certain passwords a fingerprint may be enough, but I wouldn't want to secure my PayPal or bank accounts with it. As an expert user, you may be able to assess the risks, but if we are talking about the average user, then providing them with this kind of convenience is exactly the wrong approach and the details why that is are rather hard to convey. You have to treat biometrics as usernames, not as passwords.

**Comment by leviem1** [comment]
@phoerious You're correct, but in KeePassXS, like most fingerprint accessible devices, you still need the original password initially to allow the fingerprinting to work at all, and you can configure timeouts for this. Simplifying the interface isn't going to change anything about the security of the program, which is what this thread is discussing. We don't want to change how the existing process works, just how it's presented to the user. And if a user is concerned enough about security to start using an open-source package manager over something like Dashlane, they're probably capable of making their own decision of what features they want enabled.

**Comment by phoerious** [comment]
The way it works is that we take your master password and store it in your keychain encrypted by your biometrics. We delete it again once you close KeePassXC or when you lock your computer. If we allowed for TouchID to stay active indefinitely, your master password would stay there indefinitely as well, only protected by your fingerprint.

**Comment by johnrichardrinehart** [comment]
@phoerious (please don't interpret as aggressive) What is your request? Remove biometrics/fingerprint scanning from KeePassXC? Keep the UI equally obfuscated?

I want to ensure this thread is on topic.

**Comment by phoerious** [comment]
TouchID is a quick-unlock feature, not a replacement for your master password. That is how most KeePass-compatible apps handle it and that is how it will stay for now. There is nothing "obfuscated", that is just how the mechanism works.

**Comment by leviem1** [comment]
Agreed, please don't view my last message as an attack, I'm also a bit confused what exactly you are trying to convey. I agree that "never typing a password again" is **not** secure enough. But most people are okay with the way their phones work (password on occasion and after restarts). Isn't that authentication flow what we're essentially going for here if fingerprints are enabled? If so, we've discussed just adding making another button appear for fingerprint authentication if it's available.

Conversely, we could also just make users press a button to "unlock" the DB without having the user enter a password *first*. When the user clicks the button and if they can use their fingerprint it should prompt them to do so, otherwise we prompt for the password. This is obviously harder than just adding a button, but I think it would simplify a lot about the interface.

**Comment by phoerious** [comment]
> But most people are okay with the way their phones work (password on occasion and after restarts).

That's how this here works as well. Password on occasion (e.g. after screen lock) and after restarts.

> Conversely, we could also just make users press a button to "unlock" the DB without having the user enter a password first.

How it looks in the UI is debatable and may be addressed by one of our future redesign PRs. I cannot really test any of that, since I don't own a TouchID-compatible device.

**Comment by johnrichardrinehart** [comment]
@phoerious Your concerns have been considered. There seems to be interest in improving the UI to facilitate the use of this feature and minimize confusion. I see no problem with this. We're all in agreement.

**Comment by haojianzong** [comment]
Is this being actively worked on? I just switch from LastPass to KeePassXC, from a new user's perspective, it took me a long way (~1 day ) until I figure out how to trigger TouchID. I understand using TouchID as replacement for password might be difficult, but we can make things easier for new users:

1. Include TouchID in FAQ on keepassxc.org (as already discussed above)
2. Add a "Unlock with TouchID" next to "OK" button when it is possible

**Comment by brainard52** [comment]
Another thing that is commonly done is to confirm the enablement of Touch ID by touching the sensor. Rather than a checkbox, a button could be used to trigger the confirmation dialogue.

**Comment by ostrolucky** [comment]
Nah, confirmation dialogue with touchid should be opened by default when touchid id available. That's how it's done in other password managers. We don't want having to click on some button every time we want to open database first and only then we can use fingerprint.

**Comment by brainard52** [comment]
> Nah, confirmation dialogue with touchid should be opened by default when touchid id available. That's how it's done in other password managers. We don't want having to click on some button every time we want to open database first and only then we can use fingerprint.

I meant that the button would be used in place of the checkbox. Clicking a button before doing Touch ID every time I open a database would be a pain, I agree.

**Comment by qauff** [comment]
These improvements are **very much needed**. I've been using KeePassXC **for 3 years now, and today** was the _first_ time I figured out how to unlock with my TouchID: Click "OK" first while `TouchID for Quick Unlock` is checked on (and of course not the first time you're opening your database after initial load of KeePassXC). I'm not sure if how long it took to figure out how to use TouchID says more about me as a user or more about KeePassXC's UX. Either way, here we are.

**Comment by matheusMoreno** [comment]
+1 for this. Just switched from BitWarden to KeePassXC because I wanted to store my passwords locally, but one of the best things about BitWarden is how you can **always** unlock the safe with TouchID, no password needed.

**Comment by chenzhengshui** [comment]
Agree, don’t press a button when you need to use Touch ID to unlock, it’s too much trouble. Strongbox is designed like this, no need to press a button first, thank you

**Comment by mikeziri** [comment]
I don't like complaining without contributing.
just want to say thanks for the implemented feature until now.

I agree it was hard to understand how it works (first unlock to get password in to memory for future touchid unlocks and press ok without password with touch id checkbox checked).
a FAQ on the website and a better gui will be more than enough.

thank you and keep up the good work.

**Comment by droidmonkey** [comment]
Closed in 2.7.0
